instrumentation/instaredigo/instaredigo.go

```diff
@@ -0,0 +1,175 @@
+func NewConn(sensor *instana.Sensor, netConn net.Conn, readTimeout, writeTimeout time.Duration) redis.Conn {
+    addr := netConn.LocalAddr().String()
+	conn := redis.NewConn(netConn, readTimeout, writeTimeout)
+	return &instaRedigoConn{conn, sensor, addr, nil}
+}
+
+/// Do sends a command to the server and returns the received reply and collect the instrumentation details. Inorder to capture the correlated span information, create a context from the parent span and pass it as an argument along with the other arguments. The Do API will retrieve the span information from the context and record the correlated span information.
+func (c *instaRedigoConn) Do(commandName string, args ...interface{}) (reply interface{}, err error) {
+	var cmdArgs []interface{}
+	ctx := context.Background()
+    //Separating the passed context and the arguments
+	for _, a := range args {
+		if _, ok := a.(context.Context); ok {
+			ctx = a.(context.Context)
+		} else {
+			cmdArgs = append(cmdArgs, a)
+		}
+	}
+	if c.prevSpan != nil {
+		if strings.ToUpper(commandName) == "EXEC" || strings.ToUpper(commandName) == "DISCARD" {
+			reply, err = c.Conn.Do(commandName, cmdArgs...)
+			if err != nil {
+			    c.prevSpan.span.SetTag("redis.error", err.Error())
+				c.prevSpan.span.LogFields(otlog.Object("error", err.Error()))
+			}
+			c.prevSpan.span.Finish()
+			c.prevSpan = nil
+		} else {
+			c.prevSpan.batchCmds = append(c.prevSpan.batchCmds, commandName)
+			c.prevSpan.span.SetTag("redis.subCommands", c.prevSpan.batchCmds)
+			reply, err = c.Conn.Do(commandName, cmdArgs...)
+			if err != nil {
+			    c.prevSpan.span.SetTag("redis.error", err.Error())
+				c.prevSpan.span.LogFields(otlog.Object("error", err.Error()))
+			}
+		}
+	} else {
+		tracer := c.sensor.Tracer()
+		opts := []ot.StartSpanOption{
+			ot.Tags{
+				"redis.connection": c.address,
+			},
+		}
+		if ps, ok := instana.SpanFromContext(ctx); ok {
+			tracer = ps.Tracer()
+			opts = append(opts, ot.ChildOf(ps.Context()))
+		}
+		span := tracer.StartSpan("redis", opts...)
+		if strings.ToUpper(commandName) == "MULTI" {
+			c.prevSpan = &prevSpan{span, nil}
+		}
+		span.SetTag("redis.command", commandName)
+		reply, err = c.Conn.Do(commandName, cmdArgs...)
+		if err != nil {
+			span.SetTag("redis.error", err.Error())
+			span.LogFields(otlog.Object("error", err.Error()))
+		}
+		if strings.ToUpper(commandName) != "MULTI" {
+			span.Finish()
+		}
+	}
+	return reply, err
+}
+
+// Send writes the command to the client's output buffer and collect the instrumentation details.Inorder to capture the correlated span information, create a context from the parent span and pass it as an argument along with the other arguments. The Send API will retrieve the span information from the context and record the correlated span information.
+func (c *instaRedigoConn) Send(commandName string, args ...interface{}) (err error) {
```

Comment:
Are the contents of this method exactly the same as `Do`? Maybe we could create a function where both methods can reuse?